syntax = "proto3";

package experimentserver;
//import "google/protobuf/timestamp.proto";


service Experimentserver {
	rpc connect_ensemble(stream EnsembleMessage) returns(stream EnsembleState);
	rpc connect_job(stream JobMessage) returns(stream JobState);
}

message EnsembleId {
  string experiment = 1;
  string ensemble   = 2;
}

message JobId {
  EnsembleId ensemble = 1;
  uint64 realization = 2;
  uint64 step = 3;
  uint64 job = 4;
}

enum RunState {
    UNKNOWN   = 0;
    STARTED   = 1;
    RUNNING   = 2;
    DONE      = 3;
    FAILED    = 4;
    CANCELLED = 5;
}

message EnsembleState {
  RunState runstate = 1;
  uint64 startTime = 2;
  uint64 endTime = 3;
}

message JobState {
  RunState runstate = 1;
  uint64 currentMemory = 2;
  uint64 maxMemory = 3;

//  google.protobuf.Timestamp startTime = 1;
  uint64 startTime = 4;
  uint64 endTime = 5;
}

message EnsembleMessage {
  oneof event {
  	EnsembleId id  = 1;
  	RunState state = 2;
  }
}

message JobMessage {
  oneof event {
  	JobId id       = 1;
  	JobState state = 2;
  }
}
